File: mark.py
import os
import numpy as np
import shutil

dirs = ['1-submissions','2-assignments','3-reviews','4-feedback','5-marks']

readfeedback = False

# 1. parse the review files
ranks = dict()
feedback = dict()
for marker in submissions:
    ranks[marker] = dict(alias=[],student=[],rank=[])
    feedbackfile = os.path.join(dirs[2],marker)
    f = open(feedbackfile,'r')
    lines = f.readlines()
    for line in lines:
        i = line.find('File name:')
        if (i>=0):
            bits = line.split()
            for bit in bits:
                j = bit.find('.txt')
                if (j>=0):
                    alias = bit.split('.txt')[0]
                    ranks[marker]['alias'].append(alias)
                    i = int(np.where(aliases == int(alias))[0])
                    student = submissions[i]
                    ranks[marker]['student'].append(student)
                    if (not student in feedback):
                        feedback[student] = []
        i = line.find('Rank:')
        if (i>=0):
            bits = line.split()
            for bit in bits:
                j = bit.find('/')
                if (j>=0):
                    ranks[marker]['rank'].append(int(bit.split('/')[0]))
        if readfeedback:
            i = line.find('----------------')
            if (i>=0):
                feedback[student].append(comments)
                readfeedback = False
            else:
                comments += line
        i = line.find('Feedback:')
        if (i>=0):
            readfeedback = True
            comments = line.replace("Feedback:","")
    f.close()

# 2. feedback files
keys = list(feedback.keys())
for key in keys:
    feedbackfile = os.path.join(dirs[3],str(key))
    f = open(feedbackfile, 'w')
    for comment in feedback[key]:
        f.write(comment)
    f.close()
    
# 3. PGF file
PGF = ''
keys = list(ranks.keys())
for key in keys:
    PGF += 'task1 ' + str(key) + ' '
    nr = len(ranks[key]['rank'])
    for i in range(nr-1):
        r = ranks[key]['rank'][i]
        PGF += ranks[key]['student'][r-1] + ' > '
    PGF += ranks[key]['student'][nr-1] + '\n'
PGFfile = os.path.join(dirs[4],'ranks.pgf')
f = open(PGFfile, 'w')
f.write(PGF)
f.close()

File: mark.R
require(PlayerRatings)

dirs <- c('1-submissions','2-assignments','3-reviews','4-feedback','5-marks')

readfeedback <- FALSE

# 1. parse the review files
ranks <- list()
feedback <- list()
for (m in submissions){
    ranks[[m]] <- list(alias=NULL,student=NULL,rank=NULL)
    feedbackfile <- paste0(dirs[3],'/',m)
    con <- file(feedbackfile, "r")
    for (line in readLines(con)){
        if (grepl('File name:',line)){
            string <- strsplit(line,'\\.')[[1]][1]
            a <- tail(strsplit(string,' ')[[1]],1)
            s <- submissions[which(aliases %in% a)]
            ranks[[m]]$alias <- c(ranks[[m]]$alias,a)
            ranks[[m]]$student <- c(ranks[[m]]$student,s)
        }
        if (grepl('Rank:',line)){
            string <- strsplit(line,'/')[[1]][1]
            r <- as.numeric(tail(strsplit(string,' ')[[1]],1))
            ranks[[m]]$rank <- c(ranks[[m]]$rank,r)
        }
        if (readfeedback){
            if (grepl('----------------',line)){
                feedback[[s]] <- paste0(feedback[[s]],comments)
                readfeedback <- FALSE
            } else {
                comments <- paste0(comments,line,'\n')
            }
        }
        if (grepl('Feedback:',line)){
            readfeedback <- TRUE
            comments <- strsplit(line,'Feedback:')[[1]][2]
        }
    }
    close(con)
}

# 2. feedback files
for (s in names(feedback)){
    feedbackfile <- paste0(dirs[4],'/',s)
    write(feedback[[s]],file=feedbackfile)
}

# 3. PGF file
PGF <- ''
for (m in names(ranks)){
    i <- order(ranks[[m]]$rank)
    part1 <- paste0(PGF,'task1 ',s,' ')
    part2 <- paste(ranks[[m]]$student[i],collapse=' > ')
    part3 <- '\n'
    PGF <- paste0(part1,part2,part3)
    write(PGF,file=paste0(dirs[5],'/ranks.pgf'))
}

# 4. Elo table
nr <- n*k*(k-1)/2
tt <- rep(0,nr)
p1 <- NULL
p2 <- NULL
score <- NULL
for (m in submissions){
    r <- ranks[[m]]
    for (i in 1:(k-1)){
        for (j in (i+1):k){
            p1 <- c(p1,r$student[i])
            p2 <- c(p2,r$student[j])
            if (r$rank[i]>r$rank[j]){
                score <- c(score,0)
            } else if (r$rank[i]<r$rank[j]){
                score <- c(score,1)
            } else {
                score <- c(score,1/2)
            }
        }
    }
}
ELO <- data.frame(time=tt,p1=p1,p2=p2,score=score,stringsAsFactors=FALSE)
rating <- elo(ELO)
write.table(rating$ratings,file=paste0(dirs[5],'/elo.txt'))
